runtime/interpreter/interpreter_expression.go

```diff
@@ -0,0 +1,669 @@
+		right := interpreter.evalExpression(expression.Right).(NumberValue)
+		return left.LessEqual(right)
+
+	case ast.OperationGreater:
+		left := interpreter.evalExpression(expression.Left).(NumberValue)
+		right := interpreter.evalExpression(expression.Right).(NumberValue)
+		return left.Greater(right)
+
+	case ast.OperationGreaterEqual:
+		left := interpreter.evalExpression(expression.Left).(NumberValue)
+		right := interpreter.evalExpression(expression.Right).(NumberValue)
+		return left.GreaterEqual(right)
+
+	case ast.OperationEqual:
+		left := interpreter.evalExpression(expression.Left)
+		right := interpreter.evalExpression(expression.Right)
+		return interpreter.testEqual(left, right)
+
+	case ast.OperationNotEqual:
+		left := interpreter.evalExpression(expression.Left)
+		right := interpreter.evalExpression(expression.Right)
+		return !interpreter.testEqual(left, right)
+
+	case ast.OperationOr:
+		// interpret the left-hand side
+		left := interpreter.evalExpression(expression.Left).(BoolValue)
+		// only interpret right-hand side if left-hand side is false
+		if left {
+			return left
+		}
+
+		// after interpreting the left-hand side,
+		// interpret the right-hand side
+		return interpreter.evalExpression(expression.Right).(BoolValue)
+
+	case ast.OperationAnd:
+		// interpret the left-hand side
+		left := interpreter.evalExpression(expression.Left).(BoolValue)
+		// only interpret right-hand side if left-hand side is true
+		if !left {
+			return left
+		}
+
+		// after interpreting the left-hand side,
+		// interpret the right-hand side
+		return interpreter.evalExpression(expression.Right).(BoolValue)
+
+	case ast.OperationNilCoalesce:
+		// interpret the left-hand side
+		left := interpreter.evalExpression(expression.Left)
+
+		// only evaluate right-hand side if left-hand side is nil
+		if some, ok := left.(*SomeValue); ok {
+			return some.Value
+		}
+
+		value := interpreter.evalExpression(expression.Right)
+
+		rightType := interpreter.Program.Elaboration.BinaryExpressionRightTypes[expression]
+		resultType := interpreter.Program.Elaboration.BinaryExpressionResultTypes[expression]
+
+		// NOTE: important to convert both any and optional
+		return interpreter.convertAndBox(value, rightType, resultType)
+	}
+
+	panic(&unsupportedOperation{
+		kind:      common.OperationKindBinary,
+		operation: expression.Operation,
+		Range:     ast.NewRangeFromPositioned(expression),
+	})
+}
+
+func (interpreter *Interpreter) testEqual(left, right Value) BoolValue {
+	left = interpreter.unbox(left)
+	right = interpreter.unbox(right)
+
+	// TODO: add support for arrays and dictionaries
+
+	switch left := left.(type) {
+	case NilValue:
+		_, ok := right.(NilValue)
+		return BoolValue(ok)
+
+	case EquatableValue:
+		// NOTE: might be NilValue
+		right, ok := right.(EquatableValue)
+		if !ok {
+			return false
+		}
+		return left.Equal(interpreter, right)
+
+	case *ArrayValue,
+		*DictionaryValue:
+		// TODO:
+		return false
```

Comment:
we might not need this, if `ArrayValue` and `DictionaryValue` implement the  `EquatableValue`, when we add it?

Reply:
Right, the TODO is for making arrays and dictionaries equatable (this also means allowing this statically)